File: sem3/map/ToyLanguageInterpreter/src/utils/ExecutionStackImpl.java
package utils;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Stack;

/**
 * Created by mirko on 12/10/2016.
 */
public class ExecutionStackImpl<S> implements ExecutionStack<S> {
    private Deque<S> stack;

    public ExecutionStackImpl() {
        stack = new ArrayDeque<S>();
    }

    @Override
    public void push(S statement) {
        stack.push(statement);
    }

    @Override
    public S pop() {
        if (isEmpty()) return null;
        return stack.pop();
    }

    @Override
    public boolean isEmpty() {
        return stack.isEmpty();
    }

    @Override
    public String toString() {
        StringBuilder string = new StringBuilder();
        string.append("ExecutionStackImpl = {");

        if (!stack.isEmpty()) {
            string.append("\n");
        }

        for (S statement : stack) {
            string.append("   " + statement.toString() + "\n");
        }

        string.append("}");
        return string.toString();
    }

    @Override
    public Iterable<S> getAll() {
        return stack;
    }
}

File: sem3/map/ToySharpLanguageInterpreter/ToySharpLanguageInterpreter/src/utils/ExecutionStackImpl.cs
using System.Collections.Generic;
using System.Linq;

namespace ToySharpLanguageInterpreter
{
	public class ExecutionStackImpl<S> : ExecutionStack<S>
	{
		private Stack<S> stack;

		public ExecutionStackImpl()
		{
			stack = new Stack<S>();
		}


		public void push(S statement)
		{
			stack.Push(statement);
		}


		public S pop()
		{
			return stack.Pop();
		}


		public bool isEmpty()
		{
			return stack.Count() == 0;
		}


		public override string ToString()
		{
			string str = "";
			str += "ExecutionStackImpl = {";

			if (stack.Count() > 0)
			{
				str += "\n";
			}

			foreach (S statement in stack)
			{
				str += "   " + statement + "\n";
			}

			str += "}";
			return str;
		}


		public List<S> getAll()
		{
			return stack.ToList();
		}
	}
}